Parse AMQP URI to pickup RabbitMQ connection properties

antonnaumov opened the pull request:
I've been using RabbitMQ three years in the raw and find quite useful to configure all things in the one URI. Please take a look at the pull request, this is the basics implementation. It possible to extends it to tune additional parameters

Commit:
Add an URI connection support. The current approach extends existing config based way to establish the connection. The ConnectionFactory itself also supports URI setup, but ConfigurationParams implementation doesn't allow easy setup the ConnectionFactory

Commit:
Add an URI connection support. Fix URI parsing logic to support cluster configuration

timcharper commented:
Great suggestion1 This is one of those areas where I'd prefer to do one, or the other.

is there a standard for specifying rabbitmq connections via a querystring? If so, does it specify a way to connect to multiple hosts?

antonnaumov commented:
In fact, I didn't find any standard specification provides additional settings. But I assume, if op-rabbit supports some settings it worth to support them through query parameters as well.
Speaking about multiple hosts I'm already done a URI parsing to make op-rabbit able to connect to the cluster. Please take a look at line #102

Commit:
Merge branch 'master' of github.com:SpinGo/op-rabbit

Commit:
resolve conflicts and separate test cases to better support after merge

timcharper commented:
I'm hesitant to go forward with this if there isn't a standard already in place. What is the downside of specifying the configuration using separate configuration properties?

@DStranger do you have any thoughts on this?

DStranger commented:
@timcharper rabbitmq has a [URI specification](https://www.rabbitmq.com/uri-spec.html), which, unfortunately, doesn't say anything about multiple hosts.
The way @antonnaumov has implemented seems to be in line with the way others do it, like [mongo](https://docs.mongodb.com/manual/reference/connection-string/), for example.

It would be nice to allow specifying connection params with an URI, although I don't think we should do one-or-the-other.

timcharper commented:
Right, and this is the risk: if we don't support both ways of doing it, then we create a painful migration for some of the users.

And, I don't think we want to support both ways of doing it. Especially if the extra configuration parameters (and multiple hosts) aren't a part of the RabbitMQ URI specification.

timcharper commented:
Oh, they do make a specification about additional config parameters:

https://www.rabbitmq.com/uri-query-parameters.html

DStranger commented:
Well, it's more of a `we have documented how the officially supported clients read URI query parameters`

timcharper commented:
Hmm well I can definitely see how supporting this scheme could help make things easier to configure for docker containers

antonnaumov commented:
Gentlemen, I'm wondering if we have any conclusion about this PR? I hesitate to invest more time to conflict resolution since the future of the feature is unknown. Also, the URI specification is something new in RabbitMQ implementation, and I would be happy to rework my implementation to support the official specification if the feature makes sense. 
@timcharper, the URI parameter parsing implementation is 100% falling back to the existing configuration, which means we do not bring any user of the library in trouble but add yet another method to configure the connection. In my opinion, the ongoing feature support doesn't require too much effort to support both configuration methods.

timcharper commented:
I've been thinking about this off and on for the last couple weeks. I like the idea. To repeat, my biggest concern is the potentially unpredictable behavior that will result with two configuration methods.

On the other hand, I definitely see the advantage of being able to configure the entire connection with a single value. When launching containers and passing configuration through environment variables, this is way easier to do (as opposed to needing to specify 12 or 15 environment variables). For this reason alone, I think this feature is worth it.

What about this?

- Go forward with the feature. Let's make the implementation match the RabbitMQ "specification" as much as it is "specified"
- Deprecate the existing RabbitMQ configuration settings in favor of the URL. Do not read connection information from both: Either all connection information is specified via the URL, or none of it is.

This is where I'm feeling like the decision boils down. Either it is better to specify the connection information via the URL, or it is better to specify it the way it is presently. If the URL is clearly better, then let's go that way. Otherwise, let's just keep things as is. Supporting both (as in, not having one of the methods be deprecated) is an extra burden to our users: more documentation, etc.

We could also just introduce a helper method to return a op-rabbit connection configuration from a URL, and not read it from the typesafe config object. This would enable developers to configure their connection via an environment variable more easily, if they wished, and just bypass the typesafe configuration altogether. I don't know. Feels complex. And un-opinionated.

timcharper commented:
You know, the more I think about this, the more I'm feeling like the future is with URLs.

I'm feeling more in favor of deprecating the config. Also, removing the artificial limitation that all of the posts for a cluster must be listening on the same port.

It would be nice to further push this so that Akka-RabbitMQ could show the configured URL, rather than a reconstructed URL based off of a cherry picked subset of parameters.

antonnaumov commented:
I'm glad to hear we are on the same page about configuration future. I'm going to aline the PR with plan you propose - rework the URL to support official specification as much as they specify and deprecate the existing configuration method. After that, we can discuss improvements if any or go deeper to Akka-RabbitMQ.

Thank you for your time and effort.

Commit:
Merge remote-tracking branch 'upstream/master'

# Conflicts:
#	core/src/main/scala/com/spingo/op_rabbit/ConnectionParams.scala

Commit:
Merge remote-tracking branch 'upstream/master'

# Conflicts:
#	core/src/main/scala/com/spingo/op_rabbit/ConnectionParams.scala

Commit:
Support AMQP connection URL https://www.rabbitmq.com/uri-query-parameters.html query parameters as much as possible for Scala/Java client

antonnaumov commented:
Please take another look. I think at the moment that's all parameters possible to support across Scala/Java clients. Overall, there are two limitations

1. `auth_mechanism` parameter - besides AMQP declare support multiple schemas support `com.rabbitmq.client.ConnectionFactory` still requires the particular `com.rabbitmq.client.SaslConfig` instance and supported `PLAIN` and `EXTERNAL` values only. So, my implementation follows this limitation and allow a customer to specify `auth_mechanism=plain|external`
1. `cacertfile`, `certfile`, `keyfile`, `verify` and `server_name_indication` parameters - it's might be enough to customize TLS across Erlang client, but it's not enough to proper `javax.net.ssl.SSLContext` initialization. That's why the parameters are not supported yet. Any ideas how to support TLS customization are welcome

timcharper commented on core/src/main/scala/com/spingo/op_rabbit/ConnectionParams.scala:
```diff
@@ -53,26 +55,24 @@ case class ConnectionParams(
 
-  def fromConfig(config: Config = RabbitConfig.connectionConfig) = {
-    val hosts = readHosts(config)
+  @deprecated(message = "The parameters configuration is deprecated if favor of the URL configuration", since = "2018-04-05")
```
This should have a version probably?

timcharper commented on core/src/main/scala/com/spingo/op_rabbit/ConnectionParams.scala:
```diff
@@ -84,4 +84,60 @@ object ConnectionParams {
+      case idx =>
+        parseUriHosts(uriAuthority.substring(idx + 1)) -> parseUserPassword(Some(uriAuthority.substring(0, idx)))
+    }
+    val params = Option(uri.getQuery).fold(Map.empty[String, String]) { query =>
```
I'm heavily in favor of "fail loud and proud". If the user specifies a parameter that is not support, we should fail. Can you make that a reality?

antonnaumov replied:
Absolutely, I was thinking to not be too strict for our users, but your proposal indeed makes sense. Follow the updates :)

timcharper commented on core/src/main/scala/com/spingo/op_rabbit/ConnectionParams.scala:
```diff
@@ -53,26 +55,24 @@ case class ConnectionParams(
     factory.setRequestedHeartbeat(requestedHeartbeat)
     factory.setSaslConfig(saslConfig)
-    sharedExecutor foreach factory.setSharedExecutor
+    sharedExecutor.foreach(factory.setSharedExecutor)
```
seems unrelated but okay :D

antonnaumov replied:
IntelliJ IDEA Scala plugin improving the style check, and I'm also improving the code style following the recommendations :)

timcharper commented on core/src/main/scala/com/spingo/op_rabbit/ConnectionParams.scala:
```diff
@@ -29,7 +31,7 @@ case class ConnectionParams(
     * hosts may not be tried.
     */
-  connectionTimeout: Int = 10000 /* 10 seconds */,
+  connectionTimeout: Int = ConnectionParams.DefaultConnectionTimeout,
```
Thank you!

timcharper reviewed (changes requested):
This looks super great! Sorry for being slow to review :'( If we fail hard for unsupported params (plus test to confirm) this is a slam dunk. Great change for the library.

antonnaumov committed:
Merge branch 'master' of github.com:SpinGo/op-rabbit

antonnaumov committed:
Replace deprecated annotation since timestamp with version

antonnaumov committed:
Rework the URI parameters parsing, add the validation for the parameter
names and switch to type-safe way to store values

antonnaumov committed:
Unit tests for parameters validation and processing added. Stick with
IllegalArgumentException instance whole logic with recover after
NumberFormatException if any to the IllegalArgumentException with precise
message

antonnaumov commented:
Validation is added, please take a look

DStranger commented on core/src/main/scala/com/spingo/op_rabbit/ConnectionParams.scala:
```diff
@@ -53,35 +55,133 @@ case class ConnectionParams(
+    if (config.hasPath(UriPramPath)) fromUri(new URI(config.getString(UriPramPath))) else fromParameters(config)
+  }
+
+  @deprecated(message = "The parameters configuration is deprecated if favor of the URL configuration", since = "2.1.0")
```
2.1.0 is the current version, this should be deprecated in the next one, right?

antonnaumov replied:
Oops, my bad

DStranger commented on core/src/main/scala/com/spingo/op_rabbit/ConnectionParams.scala:
```diff
@@ -53,35 +55,133 @@ case class ConnectionParams(
+  private val ChannelMaxParamName = "channel_max"
+  private val AuthMechanismParamName = "auth_mechanism"
+  sealed case class UriQueryParams(
+                                    connectionTimeout: Int = DefaultConnectionTimeout,
```
I think something went wrong here with the formatting =)

antonnaumov replied:
I'm with IntelliJ IDEA about the formatting. I'll try to make it better, but will not promise to make it ideal.

DStranger commented on core/src/main/scala/com/spingo/op_rabbit/ConnectionParams.scala:
```diff
@@ -53,35 +55,133 @@ case class ConnectionParams(
+    auth.fold(default) { candidate =>
+      candidate.split(":").toList match {
+        case usr :: psswd :: _ => (usr, psswd)
+        case _ => default
```
Why don't we fail here? I'd like to know right away in this case

antonnaumov replied:
Good catch

DStranger commented on core/src/main/scala/com/spingo/op_rabbit/ConnectionParams.scala:
```diff
@@ -53,35 +55,133 @@ case class ConnectionParams(
+  private val TlsProtectedScheme = "amqps"
+  private val DefaultConnectionTimeout = 10000 /* 10 seconds */
+
+  private val UriPramPath = "uri"
```
There's a typo here, I believe, and in the ones below.

DStranger commented on core/src/main/scala/com/spingo/op_rabbit/ConnectionParams.scala:
```diff
@@ -53,35 +55,133 @@ case class ConnectionParams(
+  private def parseAndValidateUriQuery(query: Option[String]): UriQueryParams = {
+    query.fold(Array.empty[String])(_.split('&')).foldLeft(UriQueryParams()) {
+      case (resp, par) =>
+        val index = par.indexOf('=')
```
I believe this is more readable:
`val Array(key, value) = par.split('=')`

DStranger commented on core/src/main/scala/com/spingo/op_rabbit/ConnectionParams.scala:
```diff
@@ -53,35 +55,133 @@ case class ConnectionParams(
+        val value = par.substring(index + 1)
+        key match {
+          case ConnectionTimeoutParamName =>
+            Try(resp.copy(connectionTimeout = value.toInt)).recover {
```
Since `Try` is not used here in the monadic way, I think this is more readable
```scala
try resp.copy(connectionTimeout = value.toInt)
catch {
  case _: NumberFormatException =>
  throw new IllegalArgumentException("something...")
}
```

antonnaumov replied:
In the core, I agree. My idea was to keep it compact, but it would be better to introduce a special function and use `try...catch` block.

DStranger reviewed (changes requested):
Nice work! Just a couple of comments.

DStranger commented:
@timcharper @antonnaumov 
Why don't we want to support both ways again? It shouldn't be too hard, I don't expect this code to be changing often.
I really like the old way of doing things, plus, URI's tend to become unreadable once you stuff more information in them.
Anyway, just a thought.

antonnaumov commented:
@timcharper thought it would be hard to maintain two configurations at the same time. Personally, I have a controversial opinion - on the one hand, my experience screams against duplicate functionality since it's error-prone. On the other hand, I agree it doesn't look like often changing functionality as well as it not that much code in there.

antonnaumov committed:
Typo fixing and code readability improvements after code review.
Introduce authentication credentials validation

antonnaumov committed:
Add test cases to verify the URI credentials parsing after changes

timcharper commented:
@DStranger to clarify, I said I think that we should avoid assembling the total configuration from both the URL, or from the current configuration fields (IE take some parameters from URL, and a handful of others from the config objects). It would be too error prone and messy to do otherwise.

We should "support both". It is the only sane upgrade path. We could consider deprecating the config scheme and move towards the URL configuration mechanism if there is not substantial reason to keep both; slight preference towards less configuration mechanisms to test, document, and maintain.

timcharper commented on core/src/main/scala/com/spingo/op_rabbit/ConnectionParams.scala:
```diff
@@ -53,35 +55,130 @@ case class ConnectionParams(
+
   private def readHosts(config: Config): Seq[String] = {
-    Try(config.getStringList("hosts").asScala).getOrElse(readCommaSeparatedHosts(config))
+    Try(config.getStringList(HostsPramPath).asScala).getOrElse(readCommaSeparatedHosts(config))
```
Did you mean `HostsParamPath`?

antonnaumov replied:
Yep :)

timcharper commented on core/src/main/scala/com/spingo/op_rabbit/ConnectionParams.scala:
```diff
@@ -53,35 +55,130 @@ case class ConnectionParams(
+          case AuthMechanismParamName =>
+            resp.copy(authMechanism = uriParam2DefaultSaslConfig(AuthMechanismParamName, value))
+          case _ =>
+            throw new IllegalArgumentException(s"The URL parameter [$key] is not supported")
```
👏 👏

timcharper commented on core/src/test/scala/com/spingo/op_rabbit/ConnectionParamsMultihostSpec.scala:
```diff
@@ -1,12 +1,15 @@
 
-class ConnectionParamsSpec extends FunSpec with Matchers {
+/**
+  * Created by Anton Naumov on 05.10.17.
```
Where did this comment come from?

antonnaumov replied:
That's some IntelliJ IDEA automatic stuff, thank you for pointing this I'll clean it up

timcharper reviewed (commented):
Thanks for your work on this! Everything looks great to me! There's just the minor issue of the typo and the (unintentional?) random comment :)

antonnaumov committed:
Cleanup IDEA autocomments

antonnaumov committed:
Yet another spelling error correction, thanks all my reviewers for their
attention

timcharper commented:
Woo! Merged! Thank you @antonnaumov !

antonnaumov commented:
Thank you @timcharper & @DStranger
